samples/parselib/src/main/cpp/wav/WavStreamReader.cpp

```diff
@@ -110,48 +110,203 @@ void WavStreamReader::positionToAudio() {
-            //__android_log_print(ANDROID_LOG_INFO, TAG, "read(%d)", framesThisRead);
-            int numFramesRead =
-                    mStream->read(readBuff, framesThisRead * sizeof(short) * numChans) /
-                    (sizeof(short) * numChans);
-            totalFramesRead += numFramesRead;
-
-            // convert
-            for (int offset = 0; offset < numFramesRead * numChans; offset++) {
-                buff[buffOffset++] = (float) readBuff[offset] / (float) 0x7FFF;
+        if (numFramesRead < framesThisRead) {
+            break; // none left
+        }
+
+        framesLeft -= framesThisRead;
+    }
+
+    return totalFramesRead;
+}
+
+int WavStreamReader::getDataFloat(float *buff, int numFrames) {
+    // __android_log_print(ANDROID_LOG_INFO, TAG, "getData(%d)", numFrames);
+
+    if (mDataChunk == nullptr || mFmtChunk == nullptr) {
+        return 0;
+    }
+
+    int numFramesRead = 0;
+    switch (mFmtChunk->mSampleSize) {
+        case 8:
+            numFramesRead = getDataFloat_PCM8(buff, numFrames);
+            break;
+
+        case 16:
+            numFramesRead = getDataFloat_PCM16(buff, numFrames);
+            break;
+
+        case 24:
+            if (mFmtChunk->mEncodingId == WavFmtChunkHeader::ENCODING_PCM) {
+                numFramesRead = getDataFloat_PCM24(buff, numFrames);
+            } else {
+                __android_log_print(ANDROID_LOG_INFO, TAG, "invalid encoding:%d mSampleSize:%d",
+                                    mFmtChunk->mEncodingId, mFmtChunk->mSampleSize);
             }
+            break;
 
-            if (numFramesRead < framesThisRead) {
-                break; // none left
+        case 32:
+            if (mFmtChunk->mEncodingId == WavFmtChunkHeader::ENCODING_PCM) {
+                numFramesRead = getDataFloat_PCM32(buff, numFrames);
+            } else if (mFmtChunk->mEncodingId == WavFmtChunkHeader::ENCODING_IEEE_FLOAT) {
+                numFramesRead = getDataFloat_Float32(buff, numFrames);
+            } else {
+                __android_log_print(ANDROID_LOG_INFO, TAG, "invalid encoding:%d mSampleSize:%d",
+                                    mFmtChunk->mEncodingId, mFmtChunk->mSampleSize);
             }
+            break;
 
-            framesLeft -= framesThisRead;
-        }
-        delete[] readBuff;
+        default:
+            __android_log_print(ANDROID_LOG_INFO, TAG, "invalid encoding:%d mSampleSize:%d",
+                    mFmtChunk->mEncodingId, mFmtChunk->mSampleSize);
+            return ERR_INVALID_FORMAT;
+    }
 
-        // __android_log_print(ANDROID_LOG_INFO, TAG, "  returns:%d", totalFramesRead);
-        return totalFramesRead;
+    // Zero out any unread frames
+    if (numFramesRead < numFrames) {
+        int numChannels = getNumChannels();
+        memset(buff + (numFramesRead * numChannels), 0,
+                (numFrames - numFramesRead) * mFmtChunk->mSampleSize/8 * numChannels);

```

Comment:
Target buf[] is a float array so we want sizeof(float) instead of mFmtChunk->mSampleSize/8